Userland/Libraries/LibDeviceTree/DeviceTree.cpp

```diff
@@ -7,10 +7,72 @@
 #include "DeviceTree.h"
 #include "FlattenedDeviceTree.h"
 #include <AK/NonnullOwnPtr.h>
+#include <AK/StringBuilder.h>
 #include <AK/Types.h>
 
 namespace DeviceTree {
 
+FlatPtr DeviceTreeNodeView::to_root_address(FlatPtr value) const
+{
+    // FIXME: Handle address sizes other than 32 and 64
+    // FIXME: `dma_ranges` looks very similar to this,
+    //        so maybe we can make this use this code as well
+    // ranges: [<own-addr(#address-cells)> <parent-addr(::#address-cells)> <length(#size-cells)>]
+
+    // The root node:
+    if (m_parent == nullptr)
+        return value;
+
+    auto own_ranges = get_property("ranges"sv);
+    // Note: If you hit this, you should probably call this on the parent node of your current node
+    //       This is because the most prominent user `reg` uses the parents address space
+    VERIFY(own_ranges.has_value());
+
+    if (own_ranges->size() == 0)
+        return m_parent->to_root_address(value);
+
+    auto own_address_size = get_property("#address-cells"sv);
+    VERIFY(own_address_size.has_value());
+    auto own_length_size = get_property("#size-cells"sv);
```

Comment:
I am still not a fan of calling it "length" here.
I would just call these variables own_address_cells, parent_address_cells and own_size_cells.